### PR TITLE
[TASKMGR] Implement proper process tree ending

### PR DESCRIPTION
**Pull request by Mcpg**
## Purpose
_ReactOS Task Manager has an option to shut down process trees, but it's not implemented properly (just the parent is terminated instead of the whole tree). This PR implements the missing functionality._

## Proposed changes
- Implement ShutdownProcessTree (and its helper, ShutdownProcessTreeHelper) in endproc.c and use it in ProcessPage_OnEndProcessTree.

### CODE REVIEW

File: base/applications/taskmgr/endproc.c [hunk 1]
```diff
@@ -107,6 +107,48 @@ BOOL IsCriticalProcess(HANDLE hProcess)
     return FALSE;
 }
 
+BOOL ShutdownProcessTreeHelper(HANDLE hSnapshot, HANDLE hParentProcess, DWORD dwParentPID)
+{
+    HANDLE hChildHandle;
+    PROCESSENTRY32W ProcessEntry = {0};
+    ProcessEntry.dwSize = sizeof(ProcessEntry);
+
+    if (Process32FirstW(hSnapshot, &ProcessEntry))
```
Comment (HBelusca) [inline base/applications/taskmgr/endproc.c]: Note: alternate implementation using NT API for finding parent process ID:
```c
NTSTATUS Status;
HANDLE ProcessId;
PROCESS_BASIC_INFORMATION ProcessInfo;
ULONG Length = sizeof(ProcessInfo);

/* Get the real parent's PID */

Status = NtQueryInformationProcess(TargetProcess->ProcessHandle,
                                   ProcessBasicInformation,
                                   &ProcessInfo,
                                   Length, &Length);
if (!NT_SUCCESS(Status))
{
    DPRINT1("Could not retrieve basic process info, Status = 0x%08lx\n", Status);
    return Status;
}

ProcessId = ULongToHandle(ProcessInfo.InheritedFromUniqueProcessId);
```

File: base/applications/taskmgr/endproc.c [hunk 2]
```diff
@@ -107,6 +107,48 @@ BOOL IsCriticalProcess(HANDLE hProcess)
...
+    {
+        do
+        {
+            if (ProcessEntry.th32ParentProcessID == dwParentPID)
+            {
+                hChildHandle = OpenProcess(PROCESS_TERMINATE | PROCESS_QUERY_INFORMATION,
+                                           FALSE,
+                                           ProcessEntry.th32ProcessID);
+                if (!ShutdownProcessTreeHelper(hSnapshot, hChildHandle, ProcessEntry.th32ProcessID))
```
Comment (HBelusca) [inline base/applications/taskmgr/endproc.c]: You should also check whether:
```c
    /* forbid killing system processes even if we have privileges -- sigh, windows kludge! */
    if (hProcess && IsCriticalProcess(hProcess))
...
```
happens, and if so, ignore that process and continue...

**Comment by HBelusca** [comment]
Note: related to the functionality for taskkill too, see [CORE-15998](https://jira.reactos.org/browse/CORE-15998).

File: base/applications/taskmgr/endproc.c (hunk 2 above)
Comment (Mcpg) [reply base/applications/taskmgr/endproc.c]: Sure, I'll add that check

File: base/applications/taskmgr/endproc.c [hunk 3]
```diff
@@ -107,6 +107,56 @@ BOOL IsCriticalProcess(HANDLE hProcess)
...
+    if (Process32FirstW(hSnapshot, &ProcessEntry))
+    {
+        do
+        {
+            if (ProcessEntry.th32ParentProcessID == dwParentPID)
+            {
+                hChildHandle = OpenProcess(PROCESS_TERMINATE | PROCESS_QUERY_INFORMATION,
+                                           FALSE,
+                                           ProcessEntry.th32ProcessID);
+                if (!hChildHandle || (hChildHandle && IsCriticalProcess(hChildHandle)))
```
Comment (ThFabba) [inline base/applications/taskmgr/endproc.c]: The `hChildHandle &&` is redundant.

File: base/applications/taskmgr/endproc.c [hunk 4]
```diff
@@ -107,6 +107,56 @@ BOOL IsCriticalProcess(HANDLE hProcess)
...
+                {
+                    if (hChildHandle)
+                    {
+                        CloseHandle(hChildHandle);
+                    }
+                    continue;
+                }
+                if (!ShutdownProcessTreeHelper(hSnapshot, hChildHandle, ProcessEntry.th32ProcessID))
+                {
+                    return FALSE;
```
Comment (ThFabba) [inline base/applications/taskmgr/endproc.c]: Leaking `hChildHandle`

File: base/applications/taskmgr/endproc.c (hunk 1 above)
Comment (learn-more) [reply base/applications/taskmgr/endproc.c]: We do not need to walk to the parent, we need to find all processes where our process is the parent.
aka when killing a process tree on notepad, we do not kill explorer (that started notepad) but we kill all processes started by notepad...

Comment (Mcpg) [reply base/applications/taskmgr/endproc.c]: I believe the current solution in my commit is good and there's no point in changing it to use the NT APIs (unless there's a good reason to do that).

**Comment by Mcpg** [comment]
I did use my full name in the commits.

**Pull request merged**